### ISSUE TITLE
bpf_sock: Detect when running in host netns on Kind

### ISSUE THREAD

**Issue by brb** [cilium/cilium#14956]
In bpf_sock we need to detect whether a socket belongs to a host netns (e.g. to allow a 127.0.0.1:NodePort lookup). We do that by comparing self netns cookie against a host nens cookie: https://github.com/cilium/cilium/blob/v1.9.4/bpf/bpf_sock.c#L90.

Unfortunately, this approach doesn't work when running on Kind, as cilium-agent is running in a non-host netns.

As there is no way to query a netns cookie from a userspace, one way to solve it is to write a small dummy program which runs in a separate cgroup, has attached a BPF program attached to the connect hook which pushes current process netns cookie value to a perf ring buffer. Finally, the program reads the value from the buffer, and the value is passed to bpf_sock during compile time via -D.

**Comment by johngv2** [comment]
Hi folks, I am planning take-up this issue
@pchaigno, could you please assign it to me.

**Comment by pchaigno** [comment]
@johngv2 Done. Please note this issue hasn't been flagged as `good-first-issue` so it's likely harder to tackle than our usual `good-first-issues`.

**Comment by brb** [comment]
Another suggestion by @borkmann which does not involve perf ring buffer, and thus is simpler:

> agent creates a sock, retrieves the sock cookie via getsockopt(). It then uses this sock cookie and passes it in via define. it creates the init.sh datapath. Then the bind prog for example, has a small extension which checks the sock cookie to the one of the agent socket, and if it matches, it retrieves the netns cookie and saves it in an array map with single elem. after init.sh completed that agent sock does a bind to 0.0.0.0:0 to populate the map. Then all the other hooks could use it to compare agains the 'rootns' netns cookie.

**Comment by johngv2** [comment]
Hi Folks! @brb, @borkmann... very elegant way to solve the problem, thanks for sharing. I took some time to understand some concepts around the issue, sorry about that. But I think I can raise the pr soon ;)

**Comment by aditighag** [comment]
@johngv2 Any update on the PR? Let us know if you need help.

**Comment by johngv2** [comment]
Hey @aditighag, sorry for the delay. At some point, I got stuck trying to create a **kind** cluster with more-than-one nodes. I realized the changes proposed by the commit [866969](https://github.com/cilium/cilium/pull/14951/commits/866969e3e43931c5d45b9b1a4e601f5bdf7895d4) weren't working on my machine and I tried to understand why (It would be nice to discuss it later). Anyway, I'm putting all efforts in this  issue - at least now my cluster is working fine.
